[HigherOrderOp] Fall back on all new side effects in speculate_subgraph

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104077

Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

zou3519 committed:
[HigherOrderOp] Fallback out on all new side effects in speculate_subgraph

Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by @voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[HigherOrderOp] Fallback out on all new side effects in speculate_subgraph"


Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

[ghstack-poisoned]

zou3519 committed:
Update on "[HigherOrderOp] Fallback out on all new side effects in speculate_subgraph"


Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

[ghstack-poisoned]

zou3519 committed:
Update on "[HigherOrderOp] Fall back on all new side effects in speculate_subgraph"


Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

[ghstack-poisoned]

voznesenskym commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -429,7 +876,7 @@ def false_fn(x):
         res = mod_for_compile(torch.tensor(True), torch.tensor(5))
 
-        self.assertEqual(len(backend.graphs), 2)
+        self.assertEqual(len(backend.graphs), 0)
```
hmmm why did this become 0?

zou3519 replied:
There's an append to nonlocal list in cond's true_fn. Since this is a side effect disallowed by this PR, we fall back on the entire cond to eager-mode PyTorch. The thing that is being compiled is just the cond, so there are zero graphs.

voznesenskym commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -585,7 +1024,9 @@ def f(x):
         self.assertEqual(
             dict(counters["graph_break"]),
-            {"side effects in HigherOrderOperator body": 1},
+            {
+                "HigherOrderOperator: Mutating a variable not in the current scope (SideEffects)": 1
+            },
```
very nice

voznesenskym commented on torch/_dynamo/current_scope_id.py:
```diff
@@ -0,0 +1,17 @@
+    return _current_scope_id.value
+
+
+def modify_current_scope_id(delta):
+    global _current_scope_id
+    _current_scope_id.value = current_scope_id() + delta
```
No strong opinion, but if we could do this as a context manager, you wouldn't need to invoke `modify_current_scope_id(1)` `modify_current_scope_id(-1)` you could just have this in enter/exit and it would just work.

It would also avoid someone forgetting to call this, or misaligning the delta with a != 1 or -1

zou3519 replied:
Hmm, this should be 1:1 with OutputGraph.new_subgraph_tracer, so there should be no need to actually call this outside of that function. That being said, I'll make this a context manager, it will just cause the callsite to look a bit interesting

voznesenskym commented on torch/_dynamo/side_effects.py:
```diff
@@ -15,53 +14,42 @@
+        self.source = source
 
 
-@dataclasses.dataclass
```
Where have the dataclasses gone? have you run afoul of some gross dataclass inheritance/default value / property warts?

zou3519 replied:
We now need to specify whether a mutable_local is from an existing variable that Dynamo begins to track or if the mutable local is a new variable created inside the function Dynamo is introspecting.

We use this information to assign a scope id.

This information gets specified via the constructor to MutableLocalBase, which we need to get at by actually writing constructors for these classes, so I made these regular classes so we can specify constructors for them.

voznesenskym commented on torch/_dynamo/side_effects.py:
```diff
@@ -150,8 +138,21 @@ def __contains__(self, item):
+
+        # People do things like self.dim = dim inside autograd.Function.
+        # These are benign.
+        if isinstance(item, AutogradFunctionContextVariable):
+            return True
```
Alternatively, if this expands, a more dynamo-ish pattern would be

if `item.is_side_effect_safe()`

And AutogradFunctionContextVariable can override that to return True, everything else can invoke `is_side_effect_safe` 

Not important, fine as is for now.

zou3519 replied:
Agreed -- I'll leave it for now (because I hope more cases like this don't turn up)

voznesenskym commented on torch/_dynamo/variables/base.py:
```diff
@@ -2,25 +2,81 @@
+    """
+
+    def __init__(self, typ):
+        assert typ in ("existing", "new")
```
Nit: enum?

voznesenskym commented on torch/_dynamo/variables/base.py:
```diff
@@ -2,25 +2,81 @@
+    scope_id = current_scope_id()
+    # We are allowed to mutate existing variables only if no HigherOrderOperators
+    # are involved
+    if m.scope == 0 and scope_id == 1:
+        return True
```
super nit, feel free to ignore, _is_top_level_scope() for readability

zou3519 replied:
Yeah I can make this more readable

voznesenskym reviewed (approved):
Would prefer this as a context manager where noted, but otherwise looks great

zou3519 commented:
Also, cc @tugsbayasgalan @ydwu4 if you folks wanted to take a look

zou3519 committed:
Update on "[HigherOrderOp] Fall back on all new side effects in speculate_subgraph"


Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

[ghstack-poisoned]

zou3519 committed:
Update on "[HigherOrderOp] Fall back on all new side effects in speculate_subgraph"


Fixes #103613.

A requirement for HigherOrderOperators is that after Dynamo capture, the body
function should be functional (i.e. has no observable side effects).
If the body function mutates a variable that is not local to the body, then we
that should induce a graph break.

This PR distinguish between MutableLocals created inside/outside body
and adds relevant checks. (Design originally proposed by voznesenskym.)

- We tag each mutable_local with an id that corresponds to where it came
from. The mutable_local may represent an existing object that gets
tracked by Dynamo or an object that is created while Dynamo is
introspecting.
- This id changes when we are introspecting the body of a HigherOrderOperator.
- If Dynamo wants to perform a side effect using a mutable_local, we
check its .scope field with the current scope id and raise Unsupported
in the desired case (non-local mutation inside HigherOrderOperator body)
- The id is a global thread_local variable. I can make this not a global
variable, but it just takes some engineering time to thread a number through
each of the various ways Dynamo can construct a mutable_local.

Test Plan:
- Add a bunch of new tests. Tests combinations of {global, nonlocal} x
{number, Tensor, list, object, nn.Module} and asserts that HigherOrderOp
falls back on those cases.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge
